### ISSUE TITLE
Decompound words

### ISSUE THREAD

**Issue by karussell** [komoot/photon#42]
One more normalization has to be done to improve search. E.g. Erlangerstraße will be split into "Erlanger straße". This has to be done while indexing and searching.

There is a plugin but it is GPL due to one used library, it could be less restrict but phyton code has to be ported to Java: https://github.com/jprante/elasticsearch-analysis-decompound/issues/5

For POIs there is also often the Bahnhof vs. Hauptbahnhof problem. But probably we should get a main railway station that is not named like this also important in a different country. But probably this should be handled via a different fix: #318

**Comment by hbruch** [comment]
I explored elasticsearch's hyphenation_decompounder. Currently, there exist some issues which need further work:
* ES currently requires an explicit dictionary (word_list) that must contain all subwords to be returned as tokens. Underlying lucene token filter does not, so a custom plugin that instantiates the token filter without word list would work (see [this discussion](https://discuss.elastic.co/t/why-does-hyphenation-decompounder-require-word-list/114567)).
* the hyphenation token filter returns all subwords with offsets identical to the compound word, which results in treating all subwords as synonyms in the query phase. In consequence e.g. searching 'Erlangerstraße' would treat 'erlanger' and 'strasse' as synonyms, which is not intended (see [same discussion](https://discuss.elastic.co/t/why-does-hyphenation-decompounder-require-word-list/114567)). 
* lucene's hyphenation pattern handling has a bit shifting issue (see [LUCENE-8124](https://issues.apache.org/jira/browse/LUCENE-8124)), so patterns are restricted to hyphenation markers 1 to 6. As we'd provide a custom hyphenation pattern file, so this just a minor, avoidable issue.

I'd prepare a WIP PR. The custom plugin should become an artifact on it's own, so what would you recommend: creating a github project on it's own? creating it as maven submodule?

**Comment by karussell** [comment]
Thanks, sounds good to me! I'll then close #46?

Is a separate plugin necessary because of the required internal changes? Or could we utilize it directly in photon somehow?

> creating it as maven submodule?

I would prefer this, yes.

**Comment by hbruch** [comment]
> Thanks, sounds good to me! I'll then close #46?

Yes. I'll reuse the patterns in de.txt and see if applying the decompounder only for *.de.* helps.
 
> Is a separate plugin necessary because of the required internal changes? Or could we utilize it directly in photon somehow?

Yes, internal changes are required to allow a null word_list tweak the subword offsets. To avoid JarHell exceptions, I won't patch the original code but copy and adapt. If I get things working, I'd submit them to ES/Lucene as well. But I've no idea, if hyphenation is really used in the ES/Lucene community. Current behavior looks too odd, IMHO.
